### PR TITLE
ecs_service: added deployment configuration

### LINKED ISSUE #2434: Batch mode for delegate_to and local_action (take 2)
I'll try this again.  I had suggested a feature in #2389 which was closed
before I think anyone even understood what the suggestion was.  I tried to
clarify what I meant, but it seems like comments to a closed issue go to
/dev/null.  

We have roughly 1200 servers which are currently configured with cfengine.
Cfengine is good for determining what to do based on dynamic data, but I have
wanted to find something that can handle group complexity better without
becoming unwieldy or be a resource hog.  Obviously, I'm toying around
with ansible, and trying to get it to wrap around a lot of existing stuff.

I have several times wished that delegate_to had an option to do batch
processing instead of firing off for each host in the loop and killing the
delegate server.  A batch job is supposedly possible by working around the problem as
described here:
http://jpmens.net/2012/11/03/an-action-plugin-for-ansible-to-handle-ssh-host-keys/

That solution seems less than obvious.  As a newbie working through this, the
most intuitive thing I could see would be a "batch=yes" option or something
similar, which would cause delegate_to to run once, and just send the entire
list of hosts ansible is going to loop over.  

Anyway, just a suggestion! :)

### PR DESCRIPTION
**Pull request by simplesteph**
##### ISSUE TYPE
- Feature Pull Request
##### COMPONENT NAME

ecs_service
##### ANSIBLE VERSION

```
ansible 2.3.0
  config file = 
  configured module search path = Default w/o overrides
```
##### SUMMARY

Added deployment_configuration as an optional configuration parameter for the ecs_service module, which allows to set minimum_healthy_percent and maximum_percent in both create and update functions. It doesn't do anything beyond using the existing parameters for boto3. It checks for existence of parameters and raises an exception is the arguments are not named correctly

this fixes #2434

This replaces https://github.com/ansible/ansible/pull/19316

### CODE REVIEW

**Comment by Java1Guy** [comment]
shipit

File: lib/ansible/modules/cloud/amazon/ecs_service.py [hunk 1]
```diff
@@ -180,6 +212,55 @@
 '''
 import time
 
+DEPLOYMENT_CONFIGURATION_TYPE_MAP = {
+    'maximum_percent': 'int',
+    'minimum_healthy_percent': 'int'
+}
+
+class TypeMapper:
```
Comment (bcoca) [inline lib/ansible/modules/cloud/amazon/ecs_service.py]: these would belong in the shared code in module_utils, not a show stopper at this point but consider moving them there as we already have 'decamelize' function also

Comment (simplesteph) [reply lib/ansible/modules/cloud/amazon/ecs_service.py]: @bcoca I've addressed it in my next commit. Let me know if all is good now so we can push this.

**Comment by simplesteph** [comment]
shipit

**Comment by simplesteph** [comment]
@bcoca the bot isn't picking up the ship it flag. Can you please merge?

**Comment by bcoca** [comment]
bot picked it up, but its a noop as you are already the author, the previous one by @Java1Guy is not counted as there are changes afterwards

but this one should 'count':

shipit

**Comment by anryko** [comment]
Not sure that adding automatic ```snake_dict_to_camel_dict``` parameter conversion is a good idea. Adds inconsistency with other parameters like "load_balancers"
```
        load_balancers:
          - containerName: "string"
            loadBalancerName: "string"
            containerPort: "int"
```
and converting others for consistency would introduce a breaking change.

**Comment by simplesteph** [comment]
Hi

That function is only called for the deployment configuration parameters so
it's definitely backwards compatible

Regards
Stephane

**Comment by simplesteph** [comment]
Oh I see what you're saying

I think it's just a consistency issue. Option 1 is to follow Ansible
semantics which is snake case. Option 2 is to follow boto semantics Camel
Case. Honestly for the load balancers I believe it's the wrong semantics
and we should follow Ansible semantics. But as you said it would bring a
breaking change. That's I think a concern for another PR. Happy for others
to tell us their opinion. And by the way I believe this module isn't the
only concerned by these semantics conflicts

**Comment by anryko** [comment]
Actually, it's completely fine. It will not change initial camelCase input, so it can be applied to all the other parameters (like "load_balancers") and will not brake anything. Sorry for confusion.

**Comment by simplesteph** [comment]
@ryansb @bcoca hi guys am I missing something to see this merged in, or is there a separate process? Seems like my PR is the only PR with the 'shipit' flag, are you abandoning the flag altogether? 

I'm asking because I also would like to see @anryko PR #20503 merged in but a minor rebase will have to happen after this one is merged. They're both quite important to get in, and personally speaking my PR has been opened for a few months now (back in the ansible-modules-extras repo). The reason I'm a maintainer on the ecs modules is that I wanted to help speed up the reviews and hopefully the merge process on the modules that are lagging behind the boto3 api. Let me know if I'm missing something, happy to help out :)

**Comment by mattclay** [comment]
Closing and re-opening to trigger CI.

**Comment by jctanner** [comment]
bot_status

**Pull request merged**